pep-0612.rst

```diff
@@ -0,0 +1,258 @@
+
+Motivation
+----------
+
+The existing standards for annotating higher order functions don’t give us the tools to annotate the following common decorator pattern satisfactorily:
+
+.. code-block::
+
+   from typing import Awaitable, Callable, TypeVar
+
+   TReturn = TypeVar("TReturn")
+
+   def add_logging(f: Callable[..., TReturn]) -> Callable[..., Awaitable[TReturn]]:
+       async def inner(*args: object, **kwargs: object) -> TReturn:
+           await log_to_database()
+           return f(*args, **kwargs)
+       return inner
+
+   @add_logging
+   def foo(x: int, y: str) -> int:
+       return x + 7
+
+   await foo(1, "A")
+   await foo("B", 2) # fails at runtime
+
+``add_logging``\ , a decorator which logs before each entry into the decorated function, is an instance of the Python idiom of one function passing all arguments given to it over to another function through the combination of the ``*args`` and ``**kwargs`` features in both parameters and in arguments. When one defines a function (like ``inner``\ ) that takes ``(*args, **kwargs)`` and goes on to call another function with ``(*args, **kwargs)``\ , the wrapping function can only be safely called in all of the ways that the wrapped function could be safely called. To type this decorator, we’d like to be able to place a dependency between the parameters of the callable ``f`` and the parameters of the returned function. `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_  supports dependencies between single types, as in ``def append(l: typing.List[T], e: T) -> typing.List[T]: ...``\ , but there is no existing way to do so with a complicated entity like the parameters one could pass to a function.
+
+Due to the limitations of the status quo, the ``add_logging`` example will type check but will fail at runtime. ``inner`` will pass the string “B” into ``foo``\ , which will try to add 7 to it, triggering a type error.  This was not caught by the type checker because the decorated ``foo`` was given the type ``Callable[..., Awaitable[int]]`` which is specified to do no validation on its arguments.
+
+Without the ability to define dependencies between the parameters of different callable types, there is no way, at present, to make ``add_logging`` compatible with all functions, while still preserving the enforcement of the parameters of the decorated function. 
+
+With the addition of the ``ParameterSpecification`` variables proposed by this PEP, we can rewrite the previous example in a way that keeps the flexibility of the decorator and the parameter enforcement of the decorated function.
+
+.. code-block::
+
+   from typing import Awaitable, Callable, ParameterSpecification, TypeVar
+
+   Ps = ParameterSpecification("TParams")
```

Comment:
I think this should have `"Ps"` as the call argument, not `"TParams"`.

Reply:
Corrected in https://github.com/python/peps/pull/1260